heareval/tasks/util/dataset_builder.py

```diff
@@ -0,0 +1,227 @@
+"""
+A builder class that helps to construct luigi dataset pre-processing pipelines
+"""
+import os
+from urllib.parse import urlparse
+from typing import Any, Dict, List, Union
+from types import new_class
+from functools import partial
+
+import luigi
+
+from heareval.tasks.config import DatasetConfig, PartitionedDatasetConfig
+import heareval.tasks.util.luigi as luigi_util
+
+
+class DatasetBuilder:
+    """
+    A class that helps to construct a data preprocessing pipeline for a
+    dataset.
+
+    Args:
+        task_config: the name of the task being built (will try to load the dataset
+            config file for this automatically, or a DatasetConfig file describing the
+            dataset that is being constructed.
+    """
+
+    def __init__(self, task_config: DatasetConfig):
+        if isinstance(task_config, DatasetConfig):
+            self.config = task_config
+        else:
+            raise TypeError(
+                f"task_config must be a str or a DataConfig object, "
+                f"received {type(task_config)}."
+            )
+
+        # Set the task name for all luigi WorkTasks
+        luigi_util.WorkTask.task_name = self.config.versioned_task_name
+
+    @staticmethod
+    def add_requirements(requirements: Union[luigi.Task, List, Dict], namespace: Dict):
+        """
+        Given a namespace this adds a requires() class method that returns
+        the requirments object passed in.
+
+        Args:
+            requirements: A requirements object to return in the new class method
+            namespace: namespace to add the requires method to
+        """
+        namespace["requires"] = lambda self: requirements
+
+    def build_task(
+        self,
+        base: Any,
+        name: str = None,
+        requirements: Union[luigi.Task, List, Dict] = None,
+        kwargs: Dict[str, Any] = None,
```

Comment:
kwargs might make more sense being called params since they are the parameters for the luigi task